### DISCUSSION TITLE
Customizing Escape key presses

### DISCUSSION THREAD

**Discussion by unnamedStargazer** [Exahilosys/survey#37] [Q&A] [answered]
Hello! I'm piggybacking off of [this](https://github.com/Exahilosys/survey/discussions/22) other question: how can you customize the handling of an Escape button press? I'd like the Escape button to go back to a main menu. 

Something like the following:

```py
def main_menu():
    form = {
        'Item 1': survey.widgets.Input(...),
        'Item 2': survey.widgets.Input(...),
        'Item 3': survey.widgets.Input(...),
    }

    choice = survey.routines.form('Please enter: ', form = form)

    if choice == 0: action_1()

def action_1():
    form = {
        'Item a': survey.widgets.Input(...),
        'Item b': survey.widgets.Input(...),
        'Item c': survey.widgets.Input(...),
    }

    choice = survey.routines.form('Please enter: ', form = form, callback = main_menu())
```

I tried something like this, but `callback` occurs immediately. I only want it to return to the main menu when the Escape key is pressed.

**Comment by Exahilosys** [comment, marked as answer]
Hello!

First, to clarify, you probably want to use a `routines.select` here, as a form will not return an index.

As far as esc button goes, you can disable the escape exception using `escapable = False` in any `BaseText`- and `BaseMesh`-based widget's routine (so... all of them). And then catch the respective callback and throw your own exceptions for specific control flows.

However, if all you want is to go back to the main menu, I suggest keeping the built-in escape exception and just catch it:
```py
try:
    choice = survey.routines.form('Please enter: ', form = form)
except survey.widgets.Escape:
    return main_menu()
```

Feel free to close the issue if this answers your question 🙂
